optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java

```diff
@@ -0,0 +1,318 @@
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+
+        try {
+            return (T) unmarshallerHandler.getResult();
+        } catch (JAXBException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+    }
+
+    @Override
+    public void write(T root, Writer writer) {
+        Objects.requireNonNull(root);
+        Objects.requireNonNull(writer);
+        DOMResult domResult = new DOMResult();
+        try {
+            marshaller.marshal(root, domResult);
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format(ERR_MSG_WRITE, rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+
+        formatXml(new DOMSource(domResult.getNode()), null, writer);
+    }
+
+    public void writeWithoutNamespaces(T root, Writer writer) {
+        Objects.requireNonNull(root);
+        Objects.requireNonNull(writer);
+        DOMResult domResult = new DOMResult();
+        final String errMessage = String.format(ERR_MSG_WRITE, rootClass.getName());
+        try {
+            marshaller.marshal(root, domResult);
+        } catch (JAXBException jaxbException) {
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+
+        try (InputStream xsltInputStream = getClass().getResourceAsStream("removeNamespaces.xslt")) {
+            formatXml(new DOMSource(domResult.getNode()), new StreamSource(xsltInputStream), writer);
+        } catch (IOException e) {
+            throw new OptaPlannerXmlSerializationException(errMessage, e);
+        }
+    }
+
+    private void formatXml(Source source, Source transformationTemplate, Writer writer) {
+        /*
+         * The code is not vulnerable to XXE-based attacks as it does not process any external XML nor XSL input.
+         * Should the transformerFactory be used for such purposes, it has to be appropriately secured:
+         * https://owasp.org/www-project-top-ten/OWASP_Top_Ten_2017/Top_10-2017_A4-XML_External_Entities_(XXE)
+         */
+        @SuppressWarnings({ "java:S2755", "java:S4435" })
+        TransformerFactory transformerFactory = TransformerFactory.newInstance();
+        try {
+            Transformer transformer = transformationTemplate == null ? transformerFactory.newTransformer()
+                    : transformerFactory.newTransformer(transformationTemplate);
+            // See https://stackoverflow.com/questions/46708498/jaxb-marshaller-indentation.
+            transformer.setOutputProperty(OutputKeys.INDENT, "yes");
+            transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", String.valueOf(indentation));
+            transformer.transform(source, new StreamResult(writer));
+        } catch (TransformerException transformerException) {
+            String errMessage = String.format(ERR_MSG_WRITE, rootClass.getName());
```

Comment:
Here, it didn't fail to marshall the rootClass to XML, so the error message should be different.
=> "Failed to format XML for rootClass(...)." maybe?